### ISSUE TITLE
Can't perform a React state update on an unmounted component.

### ISSUE THREAD

**Issue by jjcroftiv** [FormidableLabs/urql#493]
I have recently switched from using urql's document cache to the graph cache.  I have a table populated with the results from a query and each row has some controls for editing and deleting the record.  When using the document cache, issuing a mutation with a delete query worked as expected and the table would re-render properly without any errors.

Recently, I added a table that had a couple of different standard queries, i.e. "Open Questions" and "Resolved Questions," and the document cache couldn't resolve the deletes, and I would have to refresh the page for the record to disappear.  So, I switched to the graph cache and added mutation->updates config for the delete, which resolved the issue with the render and the record being removed from the display.

However, I now have the "Can't perform a React state update on an unmounted component." error in the console, and I can't figure out how to resolve it.  I've tried adding an abort controller to my useEffect, but this hasn't worked.  Any suggestions would be appreciated.

```
export function useDeleteOpenQuestion (keyName) {
  const [id, setId] = useState(null);
  const [result, executeMutation] = useMutation(deleteEntity);

  function handleDelete(id) {
    setId(id)
  }

  useEffect(() => {
    const abortController = new AbortController();
    if (id && !result.fetching && !result.data && !result.error) {
      var mutationParams = {};
      mutationParams[keyName] = id;
      executeMutation(mutationParams); 
    }
    if (result.error) {
      log.error(result.error)
    }
    return () => {
      abortController.abort();
    }
  });

  return handleDelete;
}
```

**Comment by JoviDeCroock** [comment]
Just a question leading up to where we want to be, is this caused by a `useMutation` or is this one from your component itself?
The code you posted will execute said effect on every render, a more elegant solution in my opinioon would be to do:

```
  function handleDelete(id) {
    executeMutation({ [keyName]: id }).then(() => do something)
  }
```

**Comment by kitten** [comment]
I'm pretty sure the issue is that we stopped guarding `setState` when we switched away from our previous `useImmediateState` primitive: https://github.com/FormidableLabs/urql/blob/master/src/hooks/useMutation.ts#L29-L35

So a mutation may lead to an update elsewhere, which may lead to a complete unmount.
If you're not expecting your current component to unmount after that mutation then you may have an issue where a `key` changes or where an entire parent tree is recreated.

On our end, I'm sure we could guard against this, but I'm not sure if it's worth it, although it'd be rather easy to readd 🤔

**Comment by jjcroftiv** [comment]
Sorry, to be more clear, the code above is just the custom hook with the abort controller that I wrote to try to work around the issue.  My first approach was to do exactly what you have in an onClick event.  I've also tried to use the Mutation component wrapping the delete button.  All with the same result.

**Comment by jjcroftiv** [comment]
@kitten Given that I'm deleting the record, I do expect it to unmount the row being deleted.  But there wasn't an issues with anything that used the document cache or any of the tables that use mutations that don't need custom configuration within graph cache.  So, I'm not sure why the custom configuration should result in different behavior.

That said, I'm also trying to work out how to put the handler and mutation in the table component where the state would be associated with a component that's not unmounted, but working out how to pass those handlers down to the buttons on the rows isn't convenient.

**Comment by kitten** [comment]
well, you could just add `rowItem.delete = () => executeMutation({ id: rowItem.id });` (imaginary code) and make sure your types (if any) match that, to conveniently have that function around on every row.

You could also just pass the handler on, but that may get tedious after a while. Or you could provide it via a small context.

**Comment by kitten** [comment]
Fixed by #496
